### PR TITLE
fix(import): close Android external document parity gaps

### PR DESCRIPTION
**Pull request by primetheus**
## Summary
- Aligns iOS external document opening with Android's installer surface for SWORD ZIP, EPUB, and app-owned TTF font files.
- Moves Files/Mail/open-in handling onto the shared import service with confirmation, ordered queued imports, and the same Settings import routing.

## Scope
- App scene URL handling in `AndBibleApp`.
- Shared external document import routing in `BibleUI`.
- Android-style TTF addon installation and startup registration in `SwordKit`/`SwordSetup`.
- Backup & Restore document import routing through the same service.
- App document type registration, localization resources, and focused unit coverage.

## Contract References
- Android parity source: Android `InstallZip` handles externally supplied ZIP, EPUB, and TTF files through installer plumbing rather than separate user-facing backup semantics.
- iOS acceptance from #185: opening a SWORD module ZIP from another app should launch AndBible and import it.
- Parity decision: iOS may use platform-specific URL/UTType plumbing, but the user-facing installer surface should not preserve platform-specific drift.

## Change Details
- Added `ExternalDocumentImportRequest` and expanded `ExternalDocumentImportService` to route ZIP, EPUB, and TTF imports by content type and extension.
- Added ZIP-to-EPUB archive fallback using structural ZIP entry inspection so EPUB packages exposed as archives route to the EPUB installer.
- Removed an unreachable duplicate EPUB reroute after archive classification was cached; the Android-aligned pre-install EPUB fallback remains unchanged.
- Added Android-style TTF installation under the SWORD root's `ttf/` directory with `AndBibleProvidesFont` addon metadata and module cache invalidation.
- Registered manually installed TTF fonts during SWORD startup so persisted font addons remain discoverable.
- Added external-open confirmation and queued import handling before mutating storage.
- Kept cancel queue advancement owned by the confirmation alert dismissal binding so canceling one prompt cannot skip the next queued import.
- Kept app-open imports file-only so unrelated URL opens cannot show import prompts or reach installers.
- Normalized provider display filenames once and reused that value for confirmation text and TTF install metadata.
- Kept generic `public.data` provider metadata selectable without letting it identify a font installer route.
- Updated Settings document import and app document types to use the shared ZIP, EPUB, and TTF support list.
- Added mirrored localization keys for font-install feedback and external-open confirmation.
- Expanded focused tests for ZIP, EPUB, TTF, unsupported documents, non-file URLs, generic provider data, provider filename trimming, ordered multi-document imports, installer failures, ZIP EPUB fallback, cached ZIP detection, unreadable TTF sources, and TTF addon config output.

## Validation Evidence
- `git diff --check`
- `plutil -lint AndBible/Info.plist AndBible.xcodeproj/project.pbxproj`
- `find AndBible Localizations -path '*/Localizable.strings' -exec plutil -lint {} +`
- `swift build --target SwordKit`
- `python3 scripts/check_settings_localization_guardrails.py`
- `python3 scripts/check_repo_standards.py all --base-ref origin/main --head-ref HEAD --all-files`
- `DEVELOPER_DIR=/Applications/Xcode.app/Contents/Developer xcodebuild test -scheme AndBible -destination 'platform=iOS Simulator,id=C52691F5-C1FB-4DC0-A4A1-00D993F9CD33'` with 13 focused external import/font test cases
- Direct Android source comparison against `InstallZip.kt` and `TtfBook.kt` for ZIP-to-EPUB fallback and TTF addon metadata behavior
- `npx gitnexus analyze`, then GitNexus `detect_changes` on the staged diffs: low risk, no affected execution flows reported
- GitNexus `detect_changes` on the import-service review-fix diff: low risk across 3 changed files, no affected indexed execution flows
- GitNexus `detect_changes` on the cancel-queue review-fix diff: low risk across 1 changed file, no affected indexed execution flows
- GitNexus `detect_changes` on the unreachable-fallback review-fix diff: low risk across 1 changed file, no mapped symbol/process impact

## Risks / Follow-ups
- Imported TTF files are app-owned reader resources, matching Android's And Bible addon metadata; this does not attempt iOS system font registration.
- The ZIP EPUB classifier only identifies EPUB package markers and otherwise leaves ZIPs on the SWORD module installer path.
- Full CI shards are expected to validate the broader suite after push; local validation focused on the changed importer, localization, and TTF persistence contracts.

Closes #185

### CODE REVIEW

File: Sources/BibleUI/Sources/BibleUI/Settings/ImportExportView.swift [hunk 1]
```diff
@@ -1028,26 +1023,21 @@ public struct ImportExportView: View {
...
-                format: String(localized: "installed_module_%@"),
-                moduleName
-            )
-        } catch {
-            statusMessage = localizedErrorMessage(error)
-        }
-        isInstallingModule = false
+        let result = ExternalDocumentImportService().importDocument(at: url)
+        statusMessage = result.feedbackMessage
+        isInstallingDocument = false
```
[the remark this answers is not in the document]
Comment (primetheus) [reply Sources/BibleUI/Sources/BibleUI/Settings/ImportExportView.swift]: This had merit. The review summary was stale after the TTF parity update, but this specific Settings-path finding was still valid: `installSupportedDocument` was running the shared importer synchronously on the main actor while the app-scene import path already used detached work. Addressed in af860d3 by yielding so the install state can render, running `ExternalDocumentImportService` in a user-initiated detached task, and publishing feedback back on the main actor. Focused external import simulator tests and repo standards passed locally.

File: Sources/BibleUI/Sources/BibleUI/Shared/ExternalDocumentImportService.swift [hunk 2]
```diff
@@ -0,0 +1,599 @@
...
+        do {
+            let moduleName = try moduleInstaller(url)
+            return .installedModule(name: moduleName)
+        } catch {
+            if epubArchiveDetector(url) {
+                return installEpub(at: url)
+            }
+            return .failed(message: error.localizedDescription)
+        }
+    }
```
[the remark this answers is not in the document]
Comment (primetheus) [reply Sources/BibleUI/Sources/BibleUI/Shared/ExternalDocumentImportService.swift:278]: This had merit. The detector result is stable for a single import attempt, and the default detector reads ZIP metadata from disk. Addressed in 8a6a237 by caching the detector result before module install and reusing it in the error path. Added a focused regression test that verifies a failed module install does not reinspect the archive.

File: AndBible/AndBibleApp.swift [hunk 3]
```diff
@@ -436,10 +447,142 @@ struct AndBibleApp: App {
...
+    private func performExternalDocumentImport(_ request: ExternalDocumentImportRequest) {
+        isImportingExternalDocument = true
+        externalDocumentImportMessage = nil
+        Task {
+            let result = await Task.detached(priority: .userInitiated) {
+                ExternalDocumentImportService().importDocument(request)
+            }.value
+            isImportingExternalDocument = false
+            externalDocumentImportMessage = result.feedbackMessage
+        }
```
[the remark this answers is not in the document]
Comment (primetheus) [reply AndBible/AndBibleApp.swift]: Partial merit. The task was created from a main-actor method, but making that actor contract explicit is the more maintainable version and avoids future ambiguity. Addressed in 8a6a237 by using an explicitly main-actor task while keeping installer work in the detached task. Focused import tests and standards passed locally.

File: AndBible/AndBibleApp.swift [hunk 4]
```diff
@@ -436,10 +447,142 @@ struct AndBibleApp: App {
...
+            ?? request.url.lastPathComponent
+            .trimmingCharacters(in: .whitespacesAndNewlines)
+        let displayName = fileName.isEmpty ? "?" : fileName
+        return String(
+            format: String(
+                localized: "install_do_you_want",
+                defaultValue: "Do you want to install %@?"
+            ),
+            displayName
+        )
```
[the remark this answers is not in the document]
Comment (primetheus) [reply AndBible/AndBibleApp.swift:583]: This had merit. Addressed in 8a6a237 by adding install_do_you_want to every mirrored AndBible and Localizations Localizable.strings file. The value follows the current repo pattern for newly introduced parity strings where untranslated locales carry the English fallback until translated upstream.

File: Sources/BibleUI/Sources/BibleUI/Shared/ExternalDocumentImportService.swift [hunk 5]
```diff
@@ -0,0 +1,599 @@
...
+        case .installedEpub(let title):
+            return String(
+                format: String(localized: "installed_epub_%@", defaultValue: "Installed EPUB: %@"),
+                title
+            )
+        case .installedFont(let name):
+            return String(
+                format: String(localized: "installed_font_%@", defaultValue: "Installed font: %@"),
+                name
+            )
```
[the remark this answers is not in the document]
Comment (primetheus) [reply Sources/BibleUI/Sources/BibleUI/Shared/ExternalDocumentImportService.swift:489]: This had merit. Addressed in 8a6a237 by adding installed_font_%@ to every mirrored AndBible and Localizations Localizable.strings file, matching the existing module and EPUB install feedback pattern. String plutil lint and localization guardrails passed locally.

File: Sources/SwordKit/Sources/SwordKit/TtfFontRepository.swift [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+                guard fileManager.isWritableFile(atPath: destination.path) else {
+                    throw TtfFontRepositoryError.cantWrite(fileName)
+                }
+                try fileManager.removeItem(at: destination)
+            }
+            do {
+                try fileManager.copyItem(at: url, to: destination)
+            } catch {
+                throw TtfFontRepositoryError.cantWrite(fileName)
+            }
```
[the remark this answers is not in the document]
Comment (primetheus) [reply Sources/SwordKit/Sources/SwordKit/TtfFontRepository.swift:137]: This had merit. Addressed in 8a6a237 by rechecking source readability after copy failure and returning the read-error path when the provider source is missing or inaccessible. Added a focused TTF repository test that verifies an unreadable source reports cantRead rather than cantWrite.

File: Sources/SwordKit/Sources/SwordKit/TtfFontRepository.swift [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+
+/**
+ Errors produced while installing an app-owned TTF font.
+
+ These mirror Android's install failure categories at the service boundary while carrying enough
+ detail for localized import feedback and tests.
+ */
+public enum TtfFontRepositoryError: LocalizedError, Sendable {
+    /// The selected file is not a readable `.ttf` file.
+    case invalidFont(String)
```
[the remark this answers is not in the document]
Comment (primetheus) [reply Sources/SwordKit/Sources/SwordKit/TtfFontRepository.swift:47]: This had merit. The code checks the sanitized filename extension, not font content readability. Addressed in 8a6a237 by changing the enum case documentation to say the selected file name does not identify a .ttf font.

File: AndBible/AndBibleApp.swift [hunk 8]
```diff
@@ -436,10 +447,142 @@ struct AndBibleApp: App {
...
+     - Side effects:
+       - queues the request when another external import prompt or result is active
+       - presents Android-style confirmation before installation
+       - later performs module/EPUB/TTF installation in a user-initiated task
+     - Failure modes: Unsupported files and installer failures are converted to localized feedback
+       by `ExternalDocumentImportService`.
+     */
+    @MainActor
+    private func handleExternalDocumentURL(_ url: URL) {
+        let request = ExternalDocumentImportRequest(url: url)
```
[the remark this answers is not in the document]
Comment (primetheus) [reply AndBible/AndBibleApp.swift:518]: Has merit. Addressed in 885ca64 by keeping the app-level external import queue file-only: non-file .onOpenURL values are ignored before any confirmation prompt is shown. I also added a service-level file URL guard so direct callers cannot route non-file URLs into installers.

File: AndBible/AndBibleApp.swift [hunk 9]
```diff
@@ -436,10 +447,142 @@ struct AndBibleApp: App {
...
+
+     - Parameter request: Pending external document request.
+     - Returns: Localized confirmation text naming the selected file.
+     - Side effects: none.
+     */
+    private func externalDocumentImportConfirmationMessage(for request: ExternalDocumentImportRequest) -> String {
+        let fileName = request.suggestedFileName
+            ?? request.url.lastPathComponent
+            .trimmingCharacters(in: .whitespacesAndNewlines)
+        let displayName = fileName.isEmpty ? "?" : fileName
```
[the remark this answers is not in the document]
Comment (primetheus) [reply AndBible/AndBibleApp.swift]: Has merit. Addressed in 885ca64 by normalizing ExternalDocumentImportRequest display filenames once after provider/URL fallback, then using that normalized value for both the confirmation prompt and TTF install metadata. Added a regression test for trimmed provider display names.

File: Sources/BibleUI/Sources/BibleUI/Shared/ExternalDocumentImportService.swift [hunk 10]
```diff
@@ -0,0 +1,600 @@
...
+
+    /// Content types accepted by the Android-parity documents importer.
+    public static var supportedContentTypes: [UTType] {
+        [.zip, .epub, trueTypeFontType, .data]
+    }
+
+    /// Dynamic UTType for TrueType fonts, with `.data` as a conservative fallback.
+    private static var trueTypeFontType: UTType {
+        UTType("public.truetype-ttf-font") ?? UTType(filenameExtension: "ttf") ?? .data
+    }
```
[the remark this answers is not in the document]
Comment (primetheus) [reply Sources/BibleUI/Sources/BibleUI/Shared/ExternalDocumentImportService.swift]: Has merit. Addressed in 885ca64 by replacing the generic public.data TTF fallback with an app-specific dynamic UTType, so generic provider metadata no longer conforms to the font type. Added regression coverage that public.data stays selectable but does not route to the font installer.

File: AndBible/AndBibleApp.swift [hunk 11]
```diff
@@ -436,10 +447,143 @@ struct AndBibleApp: App {
...
+                presenting: pendingExternalDocumentImport
+            ) { request in
+                Button(String(localized: "ok")) {
+                    pendingExternalDocumentImport = nil
+                    performExternalDocumentImport(request)
+                }
+                Button(String(localized: "cancel"), role: .cancel) {
+                    pendingExternalDocumentImport = nil
+                    showNextPendingExternalDocumentImportIfNeeded()
+                }
```
[the remark this answers is not in the document]
Comment (primetheus) [reply AndBible/AndBibleApp.swift:469]: Has merit. Addressed in dba0387 by making the confirmation alert dismissal binding the single owner for cancel queue advancement. The Cancel button no longer clears/advances the queue directly, so dismissing one prompt cannot advance once in the action and then clear the next queued request in the binding setter.

File: Sources/BibleUI/Sources/BibleUI/Shared/ExternalDocumentImportService.swift [hunk 12]
```diff
@@ -0,0 +1,638 @@
...
+            return installEpub(at: url)
+        }
+        do {
+            let moduleName = try moduleInstaller(url)
+            return .installedModule(name: moduleName)
+        } catch {
+            if isEpubArchive {
+                return installEpub(at: url)
+            }
+            return .failed(message: error.localizedDescription)
```
[the remark this answers is not in the document]
Comment (primetheus) [reply Sources/BibleUI/Sources/BibleUI/Shared/ExternalDocumentImportService.swift:276]: Has merit. Addressed in 333dbff by removing the unreachable catch-time EPUB reroute. I checked Android before changing this: Android InstallZip routes EPUB-looking ZIPs through the ZIP validation EpubFile exception path before module extraction, and iOS still preserves that behavior via the pre-install archive classifier. The removed branch was only a stale duplicate after caching the classifier result.

File: Sources/SwordKit/Sources/SwordKit/TtfFontRepository.swift [hunk 13]
```diff
@@ -0,0 +1,332 @@
...
+     Builds stable addon metadata for one TTF filename.
+
+     - Parameter fileName: Sanitized filename stored in the `ttf` directory.
+     - Returns: Font metadata with an Android-shaped `TTF_` module name.
+     */
+    private func installedFontMetadata(forFileName fileName: String) -> InstalledTtfFont {
+        let fontName = sanitizedConfigValue((fileName as NSString).deletingPathExtension)
+        let moduleName = "TTF_" + sanitizedModuleSuffix(fontName)
+        return InstalledTtfFont(fontName: fontName, moduleName: moduleName, fileName: fileName)
+    }
```
[the remark this answers is not in the document]
Comment (primetheus) [reply Sources/SwordKit/Sources/SwordKit/TtfFontRepository.swift:219]: I do not think this has merit as stated. I validated the iOS filename behavior rather than assuming: `NSString.deletingPathExtension` leaves `.ttf` as `.ttf`, so the cited example does not produce an empty fontName in this implementation. I also checked Android `TtfBook.kt`: Android derives `fontName` directly from `file.nameWithoutExtension` and does not add a stable placeholder fallback. Adding one on iOS here would introduce a new behavior that is not Android parity, so I left the code unchanged.

File: Sources/BibleUI/Sources/BibleUI/Shared/ExternalDocumentImportService.swift [hunk 14]
```diff
@@ -0,0 +1,635 @@
...
+     - Returns: `true` when central-directory or local-header metadata identifies EPUB content.
+     - Side effects: Reads ZIP metadata from `url`.
+     - Failure modes: Malformed or unreadable archives return `false`.
+     */
+    func isEpubArchive(_ url: URL) -> Bool {
+        guard let data = try? Data(contentsOf: url, options: [.mappedIfSafe]) else {
+            return false
+        }
+        let names = centralDirectoryEntryNames(in: data)
+        let entryNames = names.isEmpty ? localHeaderEntryNames(in: data) : names
```
[the remark this answers is not in the document]
Comment (primetheus) [reply Sources/BibleUI/Sources/BibleUI/Shared/ExternalDocumentImportService.swift]: Merit: yes on the memory/latency concern, but not on treating large archives as non-EPUB. Android InstallZip scans ZIP entries and reroutes when it sees META-INF/container.xml; it does not impose a large-file false negative.

Addressed in 8fbf148 by moving name-only ZIP metadata reading into BibleCore ZipArchiveReader.entryNames(inArchiveAt:). The iOS importer now reads only the EOCD search window plus central-directory headers/names one entry at a time, so large SWORD ZIP payloads are not memory-mapped or copied just to classify the archive. This preserves Android ZIP-to-EPUB parity without a size cutoff.

Also added a production-detector regression test using a real ZIP-looking EPUB fixture. Local validation passed: git diff --check, docblock guardrails, focused external import routing tests, and ZIP reader central-directory edge-case tests.

File: Sources/SwordKit/Sources/SwordKit/TtfFontRepository.swift [hunk 15]
```diff
@@ -0,0 +1,332 @@
...
+
+        """
+        let configURL = modsDirectory
+            .appendingPathComponent(font.moduleName.lowercased(), isDirectory: false)
+            .appendingPathExtension("conf")
+        do {
+            try config.write(to: configURL, atomically: true, encoding: .utf8)
+        } catch {
+            throw TtfFontRepositoryError.cantWrite(configURL.lastPathComponent)
+        }
```
[the remark this answers is not in the document]
Comment (primetheus) [reply Sources/SwordKit/Sources/SwordKit/TtfFontRepository.swift:248]: Merit: yes. Android only exposes the imported TTF display filename in the TTF install path; iOS writes the .conf as internal persistence plumbing, so leaking ttf_gentium.conf into the user-facing cantWrite message was misleading.

Fixed in ca65672 by throwing cantWrite with font.fileName from writeAddonConfig, and added testTtfFontRepositoryReportsFontNameWhenAddonConfigWriteFails to force the config-write failure and assert the selected TTF filename is reported. Also re-ran the adjacent TTF install/read-error tests.

File: Sources/SwordKit/Sources/SwordKit/TtfFontRepository.swift [hunk 16]
```diff
@@ -0,0 +1,333 @@
...
+        return fileManager.isReadableFile(atPath: url.path)
+    }
+
+    /**
+     Normalizes an external filename so it cannot escape the `ttf` directory.
+
+     - Parameter rawName: Provider-supplied filename.
+     - Returns: Non-empty basename ending in `.ttf` when the source name did.
+     */
```
[the remark this answers is not in the document]
Comment (primetheus) [reply Sources/SwordKit/Sources/SwordKit/TtfFontRepository.swift:291]: Merit: yes. The sanitizer behavior already had the UUID `.ttf` fallback for blank provider names, matching Android's generated display-name path, but the doc only described the source-name suffix case.

Fixed in a5a7ccf by documenting unsafe character replacement and the UUID `.ttf` fallback explicitly.

File: Sources/SwordKit/Sources/SwordKit/TtfFontRepository.swift [hunk 17]
```diff
@@ -0,0 +1,333 @@
...
+    @discardableResult
+    public func registerInstalledFonts() throws -> [InstalledTtfFont] {
+        try createInstallDirectories()
+        guard let files = try? fileManager.contentsOfDirectory(
+            at: fontsDirectory,
+            includingPropertiesForKeys: [.isRegularFileKey, .isReadableKey],
+            options: [.skipsHiddenFiles]
+        ) else {
+            return []
+        }
```
[the remark this answers is not in the document]
Comment (primetheus) [reply Sources/SwordKit/Sources/SwordKit/TtfFontRepository.swift:167]: Merit: partial. The documentation was ambiguous, but changing registerInstalledFonts() to throw on directory-listing failure would diverge from Android: addManuallyInstalledTtfBooks() returns when modulesDir/ttf is not readable and skips unreadable files during startup registration.

I kept the Android-parity behavior and clarified the doc in a5a7ccf: directory listing failures are treated as no readable TTF files, while directory creation and config write failures remain throwable.

File: Sources/BibleUI/Sources/BibleUI/Settings/ImportExportView.swift [hunk 18]
```diff
@@ -85,8 +85,8 @@ public struct ImportExportView: View {
     /// Whether a backup import is currently in progress.
     @State private var isImporting = false
 
-    /// Whether a SWORD module installation is currently in progress.
-    @State private var isInstallingModule = false
+    /// Whether a SWORD module or EPUB document installation is currently in progress.
+    @State private var isInstallingDocument = false
```
[the remark this answers is not in the document]
Comment (primetheus) [reply Sources/BibleUI/Sources/BibleUI/Settings/ImportExportView.swift]: Merit: yes. The shared document install path handles ZIP, EPUB, and TTF selections, so the old state comment was incomplete and could mislead future changes around the progress gate.

Fixed in 27615f4 by naming ZIP, EPUB, and TTF document installation explicitly. This is doc-only and keeps the existing Android-parity document import behavior unchanged.

File: Sources/SwordKit/Sources/SwordKit/TtfFontRepository.swift [hunk 19]
```diff
@@ -0,0 +1,335 @@
...
+            if CharacterSet.alphanumerics.contains(scalar)
+                || scalar == "."
+                || scalar == "-"
+                || scalar == "_" {
+                return Character(scalar)
+            }
+            return "_"
+        }
+        return String(scalars)
+    }
```
[the remark this answers is not in the document]
Comment (primetheus) [reply Sources/SwordKit/Sources/SwordKit/TtfFontRepository.swift:306]: Merit: partial. I validated the exact path behavior locally: `.` and `..` do not reach destination URL construction because `installFont` rejects sanitized names that do not end in `.ttf` before creating install directories. That also matches the Android TTF import path, which routes TTF imports through `.ttf` extension/display-name gating.

The valid part was that the sanitizer doc overclaimed that the helper alone enforced the `ttf/` containment guarantee. Fixed in a128ffc by clarifying that containment is the basename normalization plus `installFont` extension guard, and by adding `testTtfFontRepositoryRejectsSpecialPathComponentDisplayNames` to prove `.` and `..` are rejected before the `ttf` directory is created. Adjacent TTF repository tests pass.
